src/unity/toolkits/activity_classification/activity_classifier.cpp

```diff
@@ -539,33 +601,57 @@ void activity_classifier::perform_training_iteration() {
 
     
     const shared_float_array& output = results.at("output");
-    
-    cumulative_batch_accuracy += get_accuracy_per_batch(prediction_window, num_classes, output, batch);
-    
+
+    cumulative_batch_accuracy +=
+        get_accuracy_per_batch(prediction_window, num_classes, output, batch) /
+        batch.batch_info.size();
+
     ++num_batches;
 
   }
 
   float average_batch_loss = cumulative_batch_loss / num_batches;
   float average_batch_accuracy = cumulative_batch_accuracy / num_batches;
 
-  // Report progress if we have an active table printer.
-  // TODO: Report validation metrics.
-  
-  if (training_table_printer_) {
-    training_table_printer_->print_progress_row(
-        iteration_idx, iteration_idx + 1, average_batch_accuracy, average_batch_loss, progress_time());
 
+  if (validation_data_iterator_ == nullptr) {
```

Comment:
The remainder of this function is still hard to follow for me

Could we do something more like:
```
float average_val_accuracy;
float average_val_loss;
if (validation_data_iterator) {
  // Compute average_val_accuracy and average_val_loss
}

std::map<std::string, variant_type> state_updates({
  // training metrics
});
if (validation_data_iterator) {
  state_updates["validation_accuracy"] = ...;
  ...
}

if (training_table_printer_) {
  if (validation_data_iterator) {
    // Print six columns
  } else {
    // Print four columns
  }
}
```

(FWIW, I think doing everything with a single unified `if (validation_data_iterator_)` statement could make sense, but only if we refactor things to the point where the entire if-else statement fits on "one screen", without scrolling)